metapool/sample_sheet.py

```diff
@@ -1488,42 +1524,99 @@ def __init__(self, path=None):
+    # value to include all columns, name them with integers, and later
+    # truncate all columns that are entirely empty.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = AmpliconSampleSheet(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # remove all whitespace rows (drop all rows that are entirely empty)
+    df.dropna(how='all', axis=0, inplace=True)
 
-    sheet = MetagenomicSampleSheetv102(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # remove all comments rows, whether they are at the top of the file (no
+    # longer supported, technically), or not.
+    comment_rows = df.index[df[0].str.startswith("#")].tolist()
+    df = df.drop(index=comment_rows)
+    # reset the index to make it easier to post-process.
+    df.reset_index(inplace=True, drop=True)
 
-    sheet = MetagenomicSampleSheetv101(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # for simplicity's sake, assume the first row marks the [Header]
+    # column and raise an Error if not. By convention it should be, once
+    # legacy comments and whitespace rows are removed.
+    if df[0][0] != '[Header]':
+        raise ValueError("Top section is not [Header]")
 
-    sheet = MetagenomicSampleSheetv100(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # identify the beginning of the following section and remove everything
+    # from the start of the first section on down. Remove the now redundant
+    # [Header] from the top row as well.
+    next_section_start = df.index[df[0].str.startswith("[")].tolist()[1]
+    df = df.iloc[1:next_section_start]
 
-    sheet = MetagenomicSampleSheetv90(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # lastly, trim off the additional all-empty columns that are now
+    # present after the removal of the other sections.
+    df.dropna(how='all', axis=1, inplace=True)
 
-    sheet = MetatranscriptomicSampleSheetv10(sample_sheet_path)
-    if sheet.validate_and_scrub_sample_sheet(echo_msgs=False):
-        return sheet
+    # set the index to the attributes column of the sample-sheet, replacing
+    # the numeric index which now isn't needed. The dataframe will now just
+    # contain the index column and a single column named 1.
+    df.set_index(0, inplace=True)
+
+    # return the value of key '1'. This will return an immediately
+    # recognizable dictionary of key/value pairs.
+    results = df.to_dict()[1]
+
+    # conversion to dict causes SheetVersion to be wrapped in single ticks.
+    # e.g.: "'100'". These should be removed if present.
+    if 'SheetVersion' in results:
+        results['SheetVersion'] = results['SheetVersion'].replace("'", "")
+
+    return results
```

Comment:
Or is this the _sample sheet_ header? If so, should this be a member method of a sample sheet object?

Reply:
Same note as above: this is code merged down from main from a recent PR by @charles-cowart, so I will let him address.

Reply:
I needed a reliable method to parse the [Header] section to determine which SampleSheet() object to associate with a file. There's a legacy function in sample_sheet.py called _parse(), but it uses line-by-line parsing and csv module to read an entire sample-sheet into a SampleSheet() object. This is fine when there was only one KLSampleSheet() class but it becomes a chicken and egg problem when you have to read a sample-sheet to determine what class to associate with it.

I named the function _parse_header() because it's in the same space functionality-wise as the current _parse() function. The objective however is just to return a mapping of header attributes and values to load_sample_sheet() so it can do its thing.

I do have an agenda with this - ultimately if it all works out well I'd like to remove the line-by-line parsing and the csv module and the third party base SampleSheet() class that KLSampleSheet() inherits from. Instead I'd like it to be our own class that stores data in DataFrames, which the lab is much more familiar with. I actually already have the code to parse the entire file into data frames in another PR. Migrating KLSampleSheet() over entirely however is still farther down the road.

Reply:
Would a `staticmethod` of the `KLSampleSheet` baseclass work?

Reply:
It works for me. I was attempting to not expose it too early but it's been working well enough so far.